Bump to Trireme-Lib 8.x

bvandewalle opened the pull request:
This PR starts the work to implement the changers for Trireme-Lib 8.x.

bvandewalle commented:
This PR is now ready. @mheese , would love to have your review on this.

bvandewalle commented:
Before merging, the Gopkg.toml needs to point to master `trireme-lib` and master `trireme-csr` and `trireme-stats`

bvandewalle commented:
Depends on https://github.com/aporeto-inc/trireme-lib/pull/485

dstiliadis commented on Gopkg.toml:
```diff
@@ -3,21 +3,19 @@
 [[constraint]]
   name = "github.com/aporeto-inc/trireme-csr"
-  branch = "master"
+  branch = "trireme8.0"
```
We should use semver here

mheese replied:
@dstiliadis check out @bvandewalle 's comment on this
```
Before merging, the Gopkg.toml needs to point to master trireme-lib and master trireme-csr and trireme-stats
```

dstiliadis commented on Gopkg.toml:
```diff
@@ -3,21 +3,19 @@
   name = "github.com/aporeto-inc/trireme-statistics"
-  branch = "master"
+  branch = "add_collect_user_event"
 
```
same

mheese replied:
same

dstiliadis commented on Gopkg.toml:
```diff
@@ -3,21 +3,19 @@
 [[constraint]]
   name = "github.com/aporeto-inc/trireme-lib"
-  version = "^6.0.0"
+  branch  = "secrets_cast_check"
```
What is that ?

mheese replied:
same ...

@dstiliadis check out @bvandewalle 's comment on this
```
Before merging, the Gopkg.toml needs to point to master trireme-lib and master trireme-csr and trireme-stats
```

dstiliadis commented on Gopkg.toml:
```diff
@@ -3,21 +3,19 @@
   name = "github.com/aporeto-inc/netlink-go"
-  version = "~1.3.0"
+  branch = "context"
 
```
Not needed. netlink-go context is merged.

mheese replied:
cool, I've been waiting for that for some time :)

dstiliadis commented on resolver/policy.go:
```diff
@@ -94,43 +96,53 @@ func isPolicyUpdateNeeded(oldPod, newPod *api.Pod) bool {
+		err = k.controller.Enforce(ctx, puID, resolvedPolicy, runtime.(*policy.PURuntime))
+		if err != nil {
+			return fmt.Errorf("Error while creating the policy: %s", err)
+		}
```
What is the TODO issue? The casting?

bvandewalle replied:
yes exactly,  and also managing the other events than start eventually. Let me add the clarification for this.

mheese replied:
Yes please! I don't want to sound picky... but also, make it a switch statement with a default to be exhaustive and clear that you handled all cases. It makes it easier to read I think.

bvandewalle replied:
good point, let me change it to a switch. I actually don't  think we need or care about any other event from the Monitor.

dstiliadis commented on resolver/selector.go:
```diff
@@ -523,7 +523,7 @@ func generatePUPolicy(ingressKubeRules *[]networking.NetworkPolicyIngressRule, e
 
-	containerPolicy := policy.NewPUPolicy("", policy.Police, egressACLs, ingressACLs, egressRulesList, ingressRulesList, tags, tags, ips, triremeNets, excluded, nil)
+	containerPolicy := policy.NewPUPolicy("", policy.Police, egressACLs, ingressACLs, egressRulesList, ingressRulesList, tags, tags, ips, triremeNets, excluded, nil, nil, nil, nil)
 
```
Can you add minimal data structures here instead of nils? Although downstream checks, let's just be safe.

bvandewalle replied:
Good point, Will do.

mheese replied:
I tend to disagree on this. The downstream lib should always do input validation, and adding minimal data structures will make the code here rather verbose for something that should be easy to read.

That said, maybe we should change downstream to a builder pattern? The function call is rather long and I don't think anybody could remember what the 3rd last nil actually stands for unless you have implemented `NewPUPolicy` yourself.

bvandewalle replied:
the NewPUPolicy func is indeed getting convoluted. I would agree that we need to change it downstream. Will open a PR for that soon.

dstiliadis commented on resolver/selector.go:
```diff
@@ -584,7 +584,7 @@ func allowAllPolicy(tags *policy.TagStore, ips policy.ExtendedMap, triremeNets [
+	return policy.NewPUPolicy("", policy.Police, ingressACLs, egressACLs, nil, receivingRules, tags, tags, ips, triremeNets, excluded, nil, nil, nil, nil)
 }
 
 // notInfraContainerPolicy is a policy that should apply to the other containers in a pod that are not the infra container.
```
same

bvandewalle replied:
Good point, Will do.

mheese replied:
yeah, my disagreement from above still stands.

mheese commented on resolver/policy.go:
```diff
@@ -34,13 +35,14 @@ type KubernetesPolicy struct {
 
 // NewKubernetesPolicy creates a new policy engine for the Trireme package
-func NewKubernetesPolicy(kubeconfig string, nodename string, triremeNetworks []string, betaPolicies bool, egressPolicies bool) (*KubernetesPolicy, error) {
+func NewKubernetesPolicy(ctx context.Context, controller controller.TriremeController, kubeconfig string, nodename string, triremeNetworks []string, betaPolicies bool, egressPolicies bool) (*KubernetesPolicy, error) {
```
Do we really still need `betaPolicies` and `egressPolicies`? I thought we agreed to sticking to GA policies which would mean `betaPolicies=false` and `egressPolicies=true` in a way. Unless I'm totally mistaken and there is a good reason for this to be around, we should clean up the special cases imho.

bvandewalle replied:
This is taken care in the next PR

bvandewalle replied:
https://github.com/aporeto-inc/trireme-kubernetes/pull/155

mheese replied:
yes indeed, disregard my comment then! :)

mheese commented on resolver/policy.go:
```diff
@@ -223,15 +235,23 @@ func (k *KubernetesPolicy) updatePodPolicy(pod *api.Pod) error {
-	err = k.policyUpdater.UpdatePolicy(contextID, containerPolicy)
+
+	// TODO: Change this
+	err = k.controller.UpdatePolicy(context.TODO(), contextID, containerPolicy, runtime.(*policy.PURuntime))
```
do you want to change this because of the cast like above? I currently don't understand why this would need to be changed.

bvandewalle replied:
yes, I don't like the cast right now.  This has to chanbr in trireme-lib though. Let me clarify the TODO

mheese replied:
👍

mheese commented on resolver/policy.go:
```diff
@@ -347,9 +367,17 @@ func (k *KubernetesPolicy) updateNamespace(oldNS, updatedNS *api.Namespace) erro
-	err := k.updatePodPolicy(addedPod)
+	// Checking to see if the pod already appeared in Trireme. Returning directly if it didn't appear yet.
+	// TODO: Find a way to do a single lookup
+	_, err := k.cache.contextIDByPodName(addedPod.GetName(), addedPod.GetNamespace())
```
what do you mean by single lookup?

bvandewalle replied:
Ok, so My goal here is to remove the useless Error that we get when we get the even from Kubernetes API before receiving it from the monitors.

In order to do that, I hit the cache with this code to see if Trireme-Lib already registered a PU foe that namespace and name.

However in the next instruction we are going back to the cache again to extract the PURuntime. So we are hitting the cache twice, probably not the most efficient, I almost changed it but keeping it like that for now as it will make the code more messy,

mheese replied:
I see :) ... it totally makes sense, but what do you suggest can we do to resolve this in the future?

bvandewalle replied:
I will change it so that both can be done at once. Will do this in the next PR.

bvandewalle commented:
Thanks for your reviews @mheese @dstiliadis . I think I addressed most of your comments. If you are ok with the result, I will merge this so that we can rebase the next PRs on Trireme-Lib 8.x
